src/api/message_builder/input_selection/remainder.rs

```diff
@@ -0,0 +1,166 @@
+        }
+    }
+
+    // add burned tokens
+    let mut output_native_tokens: HashMap<TokenId, U256> = output_data.native_tokens;
+    // add burned native tokens as outputs, because we need to have this amount in the inputs
+    for (tokend_id, burned_amount) in melted_native_tokens {
+        match output_native_tokens.entry(tokend_id) {
+            Entry::Vacant(e) => {
+                e.insert(burned_amount);
+            }
+            Entry::Occupied(mut e) => {
+                *e.get_mut() += burned_amount;
+            }
+        }
+    }
+
+    let native_token_remainder = get_remainder_native_tokens(&input_native_tokens, &output_native_tokens);
+    // Output possible remaining tokens back to the original address
+    if remainder_amount > 0 {
+        let remainder_addr = get_remainder_address(remainder_address, inputs)?;
+
+        let mut remainder_output_builder = BasicOutputBuilder::new_with_amount(remainder_amount)?
+            .add_unlock_condition(UnlockCondition::Address(AddressUnlockCondition::new(remainder_addr)));
+        if let Some(remainder_native_tokens) = native_token_remainder {
+            for (token_id, amount) in remainder_native_tokens {
+                remainder_output_builder =
+                    remainder_output_builder.add_native_token(NativeToken::new(token_id, amount)?);
+            }
+        }
+        let remainder = Output::Basic(remainder_output_builder.finish()?);
+        // Check if output has enough amount to cover the storage deposit
+        remainder.verify_storage_deposit(byte_cost_config)?;
+        remainder_output.replace(remainder);
+    } else {
+        // if we have remaining native tokens, but no amount left, then we can't create this transaction unless we want
+        // to burn them
+        if native_token_remainder.is_some() && !allow_burning {
+            return Err(Error::NoBalanceForNativeTokenRemainder);
+        }
+    }
+
+    Ok(remainder_output)
+}
+
+// Return provided remainder address or get an Ed25519 address from the inputs as remainder address
+// We don't want to use nft or alias addresses as remainder address, because we might can't control them later
+pub(crate) fn get_remainder_address(remainder_address: Option<Address>, inputs: &[Output]) -> Result<Address> {
+    match remainder_address {
+        Some(address) => Ok(address),
+        // get address from an input, by default we only allow ed25519 addresses as remainder, because then we're
+        // sure that the sender can access it
+        None => {
+            let mut address = None;
+            for input in inputs {
+                if let Some(unlock_conditions) = input.unlock_conditions() {
+                    for unlock_condition in unlock_conditions.iter() {
+                        if let UnlockCondition::Address(address_unlock_condition) = unlock_condition {
+                            address.replace(address_unlock_condition.address());
+                            break;
```

Comment:
this only breaks the inner loop, but you actually want to break from both loops right?

Reply:
Good catch!